balloon/src/lib.rs

```diff
@@ -0,0 +1,395 @@
+//! [dependencies]
+//! balloon-hash = "0.1"
+//! rand_core = { version = "0.6", features = ["std"] }
+//! sha2 = "0.9"
+//! ```
+//!
+//! The following example demonstrates the high-level password hashing API:
+//!
+//! ```
+//! # fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! # #[cfg(all(feature = "password-hash", feature = "std"))]
+//! # {
+//! use balloon_hash::{
+//!     password_hash::{
+//!         rand_core::OsRng,
+//!         PasswordHash, PasswordHasher, PasswordVerifier, SaltString
+//!     },
+//!     Balloon
+//! };
+//! use sha2::Sha256;
+//!
+//! let password = b"hunter42"; // Bad password; don't actually use!
+//! let salt = SaltString::generate(&mut OsRng);
+//!
+//! // Balloon with default params
+//! let balloon = Balloon::<Sha256>::default();
+//!
+//! // Hash password to PHC string ($balloon$v=1$...)
+//! let password_hash = balloon.hash_password(password, &salt)?.to_string();
+//!
+//! // Verify password against PHC string
+//! let parsed_hash = PasswordHash::new(&password_hash)?;
+//! assert!(balloon.verify_password(password, &parsed_hash).is_ok());
+//! # }
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! [Balloon]: https://en.wikipedia.org/wiki/Balloon_hashing
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![cfg_attr(docsrs, feature(doc_cfg))]
+#![doc(
+    html_logo_url = "https://raw.githubusercontent.com/RustCrypto/media/8f1a9894/logo.svg",
+    html_favicon_url = "https://raw.githubusercontent.com/RustCrypto/media/8f1a9894/logo.svg",
+    html_root_url = "https://docs.rs/balloon-hash/0.1.0"
+)]
+#![warn(rust_2018_idioms, missing_docs)]
+
+#[cfg(feature = "alloc")]
+extern crate alloc;
+
+mod error;
+mod params;
+
+pub use crate::{
+    error::{Error, Result},
+    params::Params,
+};
+use core::convert::{TryFrom, TryInto};
+use core::marker::PhantomData;
+use core::mem;
+use core::ops::Rem;
+use crypto_bigint::{ArrayDecoding, ArrayEncoding, NonZero};
+use digest::generic_array::typenum::Unsigned;
+use digest::generic_array::GenericArray;
+use digest::Digest;
+#[cfg(feature = "password-hash")]
+#[cfg_attr(docsrs, doc(cfg(feature = "password-hash")))]
+pub use password_hash::{self, PasswordHash, PasswordHasher, PasswordVerifier};
+#[cfg(all(feature = "alloc", feature = "password-hash"))]
+use password_hash::{Decimal, Ident, ParamsString, Salt};
+
+/// Balloon context.
+///
+/// This is the primary type of this crate's API, and contains the following:
+///
+/// - Default set of [`Params`] to be used
+/// - (Optional) Secret key a.k.a. "pepper" to be used
+#[derive(Clone, Default)]
+pub struct Balloon<'key, D: Digest>
+where
+    GenericArray<u8, D::OutputSize>: ArrayDecoding,
+    <GenericArray<u8, D::OutputSize> as ArrayDecoding>::Output:
+        Rem<NonZero<<GenericArray<u8, D::OutputSize> as ArrayDecoding>::Output>>,
+    <<GenericArray<u8, D::OutputSize> as ArrayDecoding>::Output as Rem<
+        NonZero<<GenericArray<u8, D::OutputSize> as ArrayDecoding>::Output>,
+    >>::Output: ArrayEncoding,
```

Comment:
This is awful and can only be fixed if we further constrain `ArrayDecoding` and/or `ArrayEncoding` in crypto-bigint.

Reply:
See https://github.com/RustCrypto/crypto-bigint/pull/18.

Reply:
RustCrypto/crypto-bigint#18 was merged removing these constraints.